---
title: 'COVID-19 guide design (Cas13a): Test 20 candidate guides'
author: "Amanda Mok"
date: "Last compiled on `r format(Sys.Date(), '%B %d, %Y')`"
output:
  html_document: default
  pdf_document: default
---

```{r setup, echo=F}
knitr::opts_chunk$set(echo=F, fig.width=12, fig.height=8, warning=F)

library(ggplot2)
library(patchwork)
library(prodlim)
library(here)

source(file.path(here(), "scripts/helper.R"))
```

```{r load_data}
num_covid19_strains <- as.numeric(strsplit(system(paste("wc -l", file.path(here(),
                                                                           "ref_data/gisaid_cov2020_alignment.txt")),
                                                  intern=T),
                                           split=" ")[[1]][1])
num_human_CoV <- as.numeric(system(paste("grep ^'>'", file.path(here(), "ref_data/human_CoV.fa"), "| wc -l"), intern=T))

coverage_id <- "PRJNA616446"
abundance_binSize <- 300
abundance <- load_coverage(coverage_id, abundance_binSize)

plusStrand <- read.table("cas13a_20nt/cas13a_results_summary.txt", header=T, stringsAsFactors=F)
```

# I. Design

+---------------------------+-------------------------------------+----------------------------------------------------+------+
| Criteria                  | Details                             | Evaluation                                         | Done |
+===========================+=====================================+====================================================+:====:+
| Sensitivity               | Activity against SARS-CoV-2,        | Compute %detected, allow for 0 or 1 mismatches     | x    |
|                           | accounting for genomic diversity    |                                                    |      |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
| Specificity               | Low activity against other          | Compute %missed, allow 2 mismatches                | x    |
|                           | human coronaviruses                 |                                                    |      |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
|                           | Low activity against off-targets    | Align to transcriptomes, allow 2 mismatches        | !    |
|                           | (human, cow, co-occuring pathogens) | (FDA requirement: <80% homology [≥4 mismatches])   |      |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
| Good Cas13a guide         | Tag/antitag sequence constraint     | Filter out target windows that follow with GUUU ;  | x    |
|                           |                                     | flag if follow with GUUN/GUNN/GNNN                 |      |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
|                           | RNA secondary structure             | High minimum folding energy of spacer ;            | x    |
|                           |                                     | low base-pairing with crRNA hairpin ;              |      |
|                           |                                     | low base-pairing in spacer seed                    |      |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
|                           | GC content                          | Flag if GC content not between 20-50%              | x    |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
| Good viral target         | RNA abundance                       | Prioritize highly abundant subgenomic regions      | x    |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
|                           | RNA secondary structure             | Evaluate folding of sliding windows along viral    | x    |
|                           |                                     | genome, compute individual base propensity to form |      |
|                           |                                     | secondary structure                                |      |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
|                           | Target different genomic regions    | Pick target windows that span viral genome         | x    | 
+---------------------------+-------------------------------------+----------------------------------------------------+------+

x : done

! : still in progress

# II. Guide selection (plus strand)

- Number of SARS-CoV-2 strains: `r num_covid19_strains`
- Number of human coronavirus genomes (list from FDA): `r num_human_CoV`
+ NC_004718.3 : SARS CoV
+ NC_019843.3 : MERS
+ NC_006213.1 : Human CoV OC43
+ NC_006577.2 : Human CoV
+ NC_002645.1 : Human CoV 229E
+ NC_005831.2 : Human CoV NL63
- Viral abundance data set: throat swab samples from `r coverage_id`

```{r sensitivity}
sensitivity_values <- rev(c(0.99, 0.995, 0.999, 0.9992, 0.9993, 0.9995, 0.9997, 1))
sensitivity <- data.frame(sensitivity_values,
                          floor((1-sensitivity_values)*num_covid19_strains),
                          sapply(sensitivity_values,
                                 function(x) {
                                   sum(plusStrand$covid19_mismatch_0 >= x*num_covid19_strains)
                                 }),
                          sapply(sensitivity_values,
                                 function(x) {
                                   sum((plusStrand$covid19_mismatch_0+plusStrand$covid19_mismatch_1) 
                                       >= x*num_covid19_strains)
                                 }))

knitr::kable(sensitivity,
             col.names=c("Sensitivity",
                         "Number of missed strains",
                         "Number of spacers with 0 mismatches",
                         "Number of spacers with 0 or 1 mismatches"))
```

### 1. Filter out spacers that interfere with crRNA hairpin structure

```{r subset_1}
subset_1 <- subset(plusStrand, has_crRNA_hairpin)
filter_1 <- "crRNA hairpin intact"
plot_diagnostic(subset_1, abundance, filter=filter_1, 
                alpha=0.5, jitter_x=0.001, jitter_y=0.05, 
                var1="sensitivity_01", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 strains detected, allowing 1 mismatch",
                var2="specificity", var2_name="Specificity", var2_desc="% human coronaviruses missed", 
                var3="crRNA_spacer_basepairs", var3_name="crRNA structure", var3_desc="# internal spacer basepairs")
```

### 2. Filter out spacers with secondary structure (>4 base-pairs)

```{r subset_2}
subset_2 <- subset(subset_1, crRNA_spacer_basepairs <= 4)
filter_2 <- paste(filter_1, "≤4 spacer base-pairs", sep=", ")
plot_diagnostic(subset_2, abundance, filter=filter_2, 
                alpha=0.5, jitter_x=0.001, jitter_y=0.05, 
                var1="sensitivity_01", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 strains detected, allowing 1 mismatch",
                var2="specificity", var2_name="Specificity", var2_desc="% human coronaviruses missed", 
                var3="crRNA_spacer_basepairs", var3_name="crRNA structure", var3_desc="# internal spacer basepairs")
```

### 3. Filter out spacers that detect other human coronaviruses

```{r subset_3}
subset_3 <- subset(subset_2, specificity==1)
filter_3 <- paste(filter_2, "100% specificity", sep=", ")
plot_diagnostic(subset_3, abundance, filter=filter_3, 
                alpha=0.5, jitter_x=0.001, jitter_y=0.05, 
                var1="sensitivity_01", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 strains detected, allowing 1 mismatch",
                var2="GC_content", var2_name="GC content", var2_desc="spacer %GC", 
                var3="crRNA_spacer_basepairs", var3_name="crRNA structure", var3_desc="# internal spacer basepairs")
```

### 4. Filter out spacers that align to the human transcriptome (allowing 2 mismatches)

```{r subset_4}
subset_4 <- subset(subset_3, match_against_hg38==0)
filter_4 <- paste(filter_3, "\nno alignments to human transcriptome", sep=", ")
plot_diagnostic(subset_4, abundance, filter=filter_4, 
                alpha=0.5, jitter_x=0.001, jitter_y=0.05, 
                var1="sensitivity_01", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 strains detected, allowing 1 mismatch",
                var2="GC_content", var2_name="GC content", var2_desc="spacer %GC", 
                var3="crRNA_spacer_basepairs", var3_name="crRNA structure", var3_desc="# internal spacer basepairs")
```

### 5. Filter out spacers with "GUUU" antitag

```{r subset_5}
subset_5 <- subset(subset_4, antitag != "GUUU")
filter_5 <- paste(filter_4, "antitag is not GUUU", sep=", ")
plot_diagnostic(subset_5, abundance, filter=filter_5, 
                alpha=0.5, jitter_x=0.001, jitter_y=0.05, 
                var1="sensitivity_01", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 strains detected, allowing 1 mismatch",
                var2="GC_content", var2_name="GC content", var2_desc="spacer %GC", 
                var3="crRNA_spacer_basepairs", var3_name="crRNA structure", var3_desc="# internal spacer basepairs")
```

### 6. Filter out spacers with <95% sensitivity

```{r subset_6}
subset_6 <- subset(subset_5, sensitivity_01 >= 0.95)
filter_6 <- paste(filter_5, "\nsensitivity ≥ 95% (allowing 1 mismatch)", sep=", ")
plot_diagnostic(subset_6, abundance, filter=filter_6, 
                alpha=0.5, jitter_x=0.001, jitter_y=0.05, 
                var1="sensitivity_01", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 strains detected, allowing 1 mismatch",
                var2="GC_content", var2_name="GC content", var2_desc="spacer %GC", 
                var3="crRNA_spacer_basepairs", var3_name="crRNA structure", var3_desc="# internal spacer basepairs")
```

## Evaluate cross-reactivity against pathogens in FDA EUA

```{r crossreactivity}
# read in offtarget scores, subset to candidate guides
alignment_cts <- read.table("cas13a_20nt/crossreactive_offtargets_DNA.txt", header=T)
alignment_cts <- subset(alignment_cts, start %in% subset_6$start)
alignment_cts <- alignment_cts[,-c(1:2)]
alignment_summary <- data.frame(target=factor(colnames(alignment_cts), levels=colnames(alignment_cts)),
                                offtarget_rate=colMeans(alignment_cts>0),
                                target_type=NA)
# specify offtarget groups
human_CoVs <- c("huCoV_229E", "SARS", "huCoV_NL63", "huCoV_OC43", "huCoV_HKU1", "MERS")
offtarget_ssRNA_plusStrand <- c("enterovirus", "rhinovirus", "parechovirus")
offtarget_ssRNA_minusStrand <- c("hMPV", paste0("parainfluenza_", 1:4), paste0("influenza_", c("A", "B")), "resp_syncytial_A")
offtarget_dsDNA <- c("adenovirus", "c_pneumoniae", "h_influenzae", "l_pneumophila", "m_tuberculosis", "s_pneumoniae", "s_pyogenes",
                     "b_pertussis", "m_pneumoniae", "p_jirovecii", "c_albicans", "p_aeruginosa", "s_epidermidis", "s_salivarius")
alignment_summary$target_type[colnames(alignment_cts) %in% human_CoVs] <- "human CoVs"
alignment_summary$target_type[colnames(alignment_cts) %in% offtarget_ssRNA_minusStrand] <- "ssRNA (-)"
alignment_summary$target_type[colnames(alignment_cts) %in% offtarget_ssRNA_plusStrand] <- "ssRNA (+)"
alignment_summary$target_type[colnames(alignment_cts) %in% offtarget_dsDNA] <- "dsDNA"
# plot crossreactivity
ggplot(alignment_summary, aes(target, offtarget_rate)) + 
  geom_col(aes(fill=target_type)) + theme_bw() + theme(axis.text.x=element_text(angle=90)) + 
  ggtitle(paste0("off-target cross-reactivity: ", "DNA"), 
          subtitle=paste0("candidate spacers (n=", nrow(alignment_cts), ")")) + 
  xlab("") + ylab("% spacers aligned with ≤4 mismatches")
```

# III. Guide selection (minus strand)

```{r minus_strand}
minusStrand <- read.table("cas13a_minusStrand_20nt/cas13a_minusStrand_results_summary.txt", header=T) # 
minusStrand <- subset(minusStrand, has_crRNA_hairpin) # n = 15938
minusStrand <- subset(minusStrand, crRNA_spacer_basepairs <= 4) # n = 8070
minusStrand <- subset(minusStrand, specificity==1) # n = 5502
minusStrand <- subset(minusStrand, match_against_hg38==0) # n = 3375
minusStrand <- subset(minusStrand, antitag != "GUUU") # n = 3345
minusStrand <- subset(minusStrand, sensitivity_01 >= 0.95) # n = 3291 
plot_diagnostic(minusStrand, abundance, filter="", 
                alpha=0.5, jitter_x=0.001, jitter_y=0.05, 
                var1="sensitivity_01", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 strains detected, allowing 1 mismatch",
                var2="GC_content", var2_name="GC content", var2_desc="spacer %GC", 
                var3="crRNA_spacer_basepairs", var3_name="crRNA structure", var3_desc="# internal spacer basepairs")
```

# IV. Merge candidate spacers (plus and minus strands)

```{r candidates}
candidates <- rbind(subset_6, minusStrand)
candidates <- candidates[,c("start", "target", "spacer", "strand")]
write.table(candidates, file="candidate_spacers.txt",
            quote=F, row.names=F, sep="\t")
```

# V. Amplicon guides

```{r amplicon_guides}
# read in alignments
amplicons <- system(paste("grep -v @", file.path(here(), "isothermal_amplification/winner_amplicons_mapped.sam"), 
                          "| cut -f 1,2,4,10"), intern=T)
amplicons <- data.frame(matrix(unlist(strsplit(amplicons, split="\t")), byrow=T, ncol=4), stringsAsFactors=F)
colnames(amplicons) <- c("name", "flag", "position", "sequence")
amplicons$position <- as.numeric(amplicons$position)

amplicon_overlap <- lapply(seq.int(amplicons),
                           function(x) {
                             amplicon_start <- amplicons$position[x]
                             amplicon_end <- amplicons$position[x] + nchar(amplicons$sequence[x])
                             spacer_stop <- amplicon_end - 20 + 1
                             tmp_subset <- subset(candidates, start >= amplicon_start & start <= spacer_stop)
                             if(nrow(tmp_subset) > 0) {
                               return(data.frame(tmp_subset, amplicon=amplicons$name[x]))
                             } else {
                               return(NULL)
                             }
                           })
amplicon_overlap <- do.call(rbind, amplicon_overlap)
# aggregate(start~strand+amplicon, data=amplicon_overlap, FUN=length)

# assign to regions of overlapping guides (≥10 nt)
amplicon_overlap <- amplicon_overlap[order(amplicon_overlap$start),]
for(x in seq.int(nrow(amplicon_overlap))) {
  if(x == 1) {
    region <- 1
  } else {
    if(amplicon_overlap$start[x] >= (amplicon_overlap$start[x-1]+10)) {
      region <- region+1
    }
  }
  amplicon_overlap$region[x] <- region
}
aggregate(start~strand+region+amplicon, data=amplicon_overlap, FUN=length)
```

# VI. Offtarget analysis: Pollard lab

```{r offtarget}
viral <- read.delim(file.path(here(), "ref_data", "offtarget_analysis", "classified_kmers_vn_hd_lt_2_20200609.tsv"), 
                    header=T, stringsAsFactors=F)
bacterial_0 <- read.delim(file.path(here(), "ref_data", "offtarget_analysis", "classified_kmers_refbac_hd.0_20200609.tsv"),
                          header=T, stringsAsFactors=F)
bacterial_1 <- read.delim(file.path(here(), "ref_data", "offtarget_analysis", "classified_kmers_refbac_hd.1_20200609.tsv"),
                          header=T, stringsAsFactors=F)


```

# VII. Ott guides

```{r ott_guides}

```