---
title: "COVID-19 guide design (Cas12): Test 20 candidate guides"
author: "Amanda Mok"
date: "Last compiled on `r format(Sys.Date(), '%B %d, %Y')`"
output: html_document
---

```{r setup, echo=F}
knitr::opts_chunk$set(echo=F, fig.width=12, fig.height=8, warning=F)

library(ggplot2)
library(patchwork)
library(prodlim)
library(here)

source(file.path(here(), "scripts/helper.R"))
```

```{r load_data}
num_covid19_strains <- as.numeric(strsplit(system(paste("wc -l", file.path(here(),
                                                                           "ref_data/gisaid_cov2020_alignment.txt")),
                                                  intern=T),
                                           split=" ")[[1]][1])
num_human_CoV <- as.numeric(system(paste("grep ^'>'", file.path(here(), "ref_data/human_CoV.fa"), "| wc -l"), intern=T))

coverage_id <- "PRJNA616446"
abundance_binSize <- 300
abundance <- load_coverage(coverage_id, abundance_binSize)

windows <- read.table("cas12_results_summary.txt", header=T, stringsAsFactors=F)
output_columns <- c("target", "spacer", "start", "antitag", "target_basepairing_prob", "match_against_bosTau9")
```

# Design

AIM: 10-20 candidate spacers, 20 nt length

+---------------------------+-------------------------------------+----------------------------------------------------+------+
| Criteria                  | Details                             | Evaluation                                         | Done |
+===========================+=====================================+====================================================+:====:+
| Sensitivity               | Activity against SARS-CoV-2,        | Compute %detected, allow for 0 or 1 mismatches     | x    |
|                           | accounting for genomic diversity    |                                                    |      |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
| Specificity               | Low activity against other          | Compute %missed, allow 2 mismatches                | x    |
|                           | human coronaviruses                 |                                                    |      |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
|                           | Low activity against off-targets    | Align to transcriptomes, allow 2 mismatches        | !    |
|                           | (human, cow, co-occuring pathogens) | (FDA requirement: <80% homology [≥4 mismatches])   |      |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
| Good Cas12 guide          | PAM sequence constraint             | Select TTTN PAM sequence ;   flag if PAM is TTTT   | x    |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
|                           | RNA secondary structure             | High minimum folding energy of spacer ;            | x    |
|                           |                                     | low base-pairing with crRNA hairpin ;              |      |
|                           |                                     | low base-pairing in spacer seed                    |      |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
|                           | GC content                          | Flag if GC content not between 20-50%              | x    |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
| Good viral target         | RNA abundance                       | Prioritize highly abundant subgenomic regions      | x    |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
|                           | RNA secondary structure             | Evaluate folding of sliding windows along viral    | x    |
|                           |                                     | genome, compute individual base propensity to form |      |
|                           |                                     | secondary structure                                |      |
+---------------------------+-------------------------------------+----------------------------------------------------+------+
|                           | Target different genomic regions    | Pick target windows that span viral genome         | x    | 
+---------------------------+-------------------------------------+----------------------------------------------------+------+

x : done

! : still in progress

# Selection of target window

- Number of SARS-CoV-2 genomes: `r num_covid19_strains`
- Number of human coronavirus genomes (list from FDA): `r num_human_CoV`
    + NC_004718.3 : SARS CoV
    + NC_019843.3 : MERS
    + NC_006213.1 : Human CoV OC43
    + NC_006577.2 : Human CoV
    + NC_002645.1 : Human CoV 229E
    + NC_005831.2 : Human CoV NL63
- Viral abundance data set: throat swab samples from `r coverage_id`

```{r sensitivity}
sensitivity_values <- rev(c(0.99, 0.995, 0.999, 0.9992, 0.9993, 0.9995, 0.9997, 1))
sensitivity <- data.frame(sensitivity_values,
                          floor((1-sensitivity_values)*num_covid19_strains),
                          sapply(sensitivity_values,
                                 function(x) {
                                   sum(windows$covid19_mismatch_0 >= x*num_covid19_strains)
                                 }),
                          sapply(sensitivity_values,
                                 function(x) {
                                   sum((windows$covid19_mismatch_0+windows$covid19_mismatch_1) 
                                       >= x*num_covid19_strains)
                                 }))

knitr::kable(sensitivity,
             col.names=c("Sensitivity",
                         "Number of missed genomes",
                         "Number of spacers with 0 mismatches",
                         "Number of spacers with 0 or 1 mismatches"))
```

### 0. Select spacers with correct PAM

```{r diagnostic_all}
plot_diagnostic(windows, abundance, 
                alpha=0.2, jitter_x=0.01, jitter_y=0.04,
                var1="sensitivity_0", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 genomes captured, 0 mismatches",
                var2="specificity", var2_name="Specificity", var2_desc="% human coronaviruses missed",
                var3="match_against_hg38", var3_name="Reactivity: hg38", var3_desc="# alignments to human genome")
```

### 1. Filter out spacers that interfere with crRNA hairpin

```{r subset_1}
subset_1 <- subset(windows, has_crRNA_hairpin)
filter_1 <- "crRNA hairpin intact"
plot_diagnostic(subset_1, abundance, filter=filter_1,
                alpha=0.2, jitter_x=0.01, jitter_y=0.04,
                var1="sensitivity_0", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 genomes captured, 0 mismatches",
                var2="specificity", var2_name="Specificity", var2_desc="% human coronaviruses missed",
                var3="match_against_hg38", var3_name="Reactivity: hg38", var3_desc="# alignments to human genome")
```

### 2. Filter out spacers with >4 internal base-paired positions

```{r subset_2}
subset_2 <- subset(subset_1, crRNA_spacer_basepairs > 4)
filter_2 <- paste(filter_1, "≤4 spacer base-pairs", sep=", ")
plot_diagnostic(subset_2, abundance, filter=filter_2,
                alpha=0.2, jitter_x=0.01, jitter_y=0.04,
                var1="sensitivity_0", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 genomes captured, 0 mismatches",
                var2="specificity", var2_name="Specificity", var2_desc="% human coronaviruses missed",
                var3="match_against_hg38", var3_name="Reactivity: hg38", var3_desc="# alignments to human genome")
```

### 3. Select spacers with 100% specificity

```{r subset_3}
subset_3 <- subset(subset_2, specificity==1)
filter_3 <- paste(filter_2, "100% specificity", sep=", ")
plot_diagnostic(subset_3, abundance, filter=filter_3,
                alpha=0.5, jitter_x=0.01, jitter_y=0.3,
                var1="sensitivity_0", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 genomes captured, 0 mismatches",
                var2="match_against_hg38", var2_name="Reactivity: hg38", var2_desc="# alignments to human genome",
                var3="GC_content", var3_name="% GC", var3_desc="Spacer GC content")
```

### 4. Filter out spacers with >10 alignments to human genome

```{r subset_4}
subset_4 <- subset(subset_3, match_against_hg38 > 10)
filter_4 <- paste(filter_3, "\n≤10 alignments to human genome", sep=", ")
plot_diagnostic(subset_4, abundance, filter=filter_4,
                alpha=0.5, jitter_x=0.005, jitter_y=0.3,
                var1="sensitivity_0", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 genomes captured, 0 mismatches",
                var2="match_against_hg38", var2_name="Reactivity: hg38", var2_desc="# alignments to human genome",
                var3="GC_content", var3_name="% GC", var3_desc="Spacer GC content")
```

### 5. Filter out PAM sequences that are TTTT

```{r subset_5}
subset_5 <- subset(subset_4, PAM != "TTTT")
filter_5 <- paste(filter_4, "PAM is not TTTT", sep=", ")
plot_diagnostic(subset_5, abundance, filter=filter_5,
                alpha=0.5, jitter_x=0.005, jitter_y=0.3,
                var1="sensitivity_0", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 genomes captured, 0 mismatches",
                var2="match_against_hg38", var2_name="Reactivity: hg38", var2_desc="# alignments to human genome",
                var3="GC_content", var3_name="% GC", var3_desc="Spacer GC content")
```

### 6. Select one spacer per overlapping window

- Window overlap: ≥10 nt 
- Spacer prioritization:
    + Fewer alignments to human genome
    + Higher sensitivity

```{r subset_6}
subset_6 <- subset_5
subset_6$region <- NA
for(x in seq.int(nrow(subset_6))) {
  if(x == 1) {
    region <- 1
  } else {
    if(subset_6$start[x] >= (subset_6$start[x-1]+10)) {
      region <- region+1
    }
  }
  subset_6$region[x] <- region
}
subset_6 <- lapply(unique(subset_6$region),
                   function(x) {
                     tmp <- subset(subset_6, region==x)
                     if(nrow(tmp > 1)) {
                       tmp <- subset(tmp, match_against_hg38==min(tmp$match_against_hg38))
                       if(nrow(tmp > 1)) {
                         tmp <- subset(tmp, sensitivity_0==max(tmp$sensitivity_0))
                         if(nrow(tmp >1)) {
                           return(tmp[1,])
                         }
                       }
                     }
                     return(tmp)
                   })
subset_6 <- do.call(rbind, subset_6)
filter_6 <- paste(filter_5, "\n≤10 nt overlap", sep=", ")
plot_diagnostic(subset_6, abundance, filter=filter_6,
                alpha=0.5, jitter_x=0.005, jitter_y=0.3,
                var1="sensitivity_0", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 genomes captured, 0 mismatches",
                var2="match_against_hg38", var2_name="Reactivity: hg38", var2_desc="# alignments to human genome",
                var3="GC_content", var3_name="% GC", var3_desc="Spacer GC content")
```

### 7. Select top 20 by sensitivity

- Allowing 0 mismatches

```{r subset_7}
subset_7 <- subset_6[order(subset_6$sensitivity_0, decreasing=T)[1:20],]
filter_7 <- paste(filter_6, "top 20 by sensitivity", sep=", ")
plot_diagnostic(subset_7, abundance, filter=filter_7,
                alpha=0.5, jitter_x=0, jitter_y=0.3,
                var1="sensitivity_0", var1_name="Sensitivity", var1_desc="% SARS-CoV-2 genomes captured, 0 mismatches",
                var2="match_against_hg38", var2_name="Reactivity: hg38", var2_desc="# alignments to human genome",
                var3="GC_content", var3_name="% GC", var3_desc="Spacer GC content")
```

# Candidate spacers: subset #7

```{r output_subset_7}
output_columns <- c("target", "spacer", "start", "PAM", "match_against_hg38")
output_7 <- subset_7[, output_columns]
rownames(output_7) <- seq.int(nrow(output_7))

knitr::kable(output_7,
             row.names=T,
             col.names=c("Viral target window",
                         "Spacer sequence",
                         "Genomic position",
                         "PAM",
                         "# alignments to human genome"))
```

```{r ivt_window, eval=F}
genome_seq <- readLines(file.path(here(), "ref_data/NC_045512v2.fa"))
genome_seq <- genome_seq[!grepl(">", genome_seq)]
genome_seq <- paste0(genome_seq, collapse="")

output_7$viral_genome_5prime <- sapply(output_7$start,
                                       function(x) {
                                         substr(genome_seq, start=x-20, stop=x-1)
                                       })
output_7$viral_genome_3prime <- sapply(output_7$start,
                                       function(x) {
                                         substr(genome_seq, start=x+20, stop=x+20+20-1)
                                       })
output_7$IVT_window <- sapply(output_7$start,
                              function(x) {
                                substr(genome_seq, start=x-20, stop=x+20+20-1)
                              })
write.csv(output_7, paste0("guide_summary_", format(Sys.time(), "%Y%m%d"), ".csv"))
```